canonical/ubuntu-advantage-client
# apt-hook: run new json hook tests on package build

## Proposed Commit Message


> apt-hook: run new json hook tests on package build
> 
> The new json hook is written in golang and has some
> go unit tests. We are now running these unit tests
> when building a deb to help prevent regressions.
> 
> While the go program itself does not require a
> particular newer version of go installed, the tests
> do. The `t.Run` function we are using requires at
> least version 1.7 to run. Because of this, we are
> adjusting our Build-Depends appropriately. We will
> now use `golang-1.10` if available (as it is in older
> releases where the default version is less than 1.7),
> and just use `golang` if not (in newer releases
> where the default version is greater than 1.7).
> 
> Fixes: #1567

> apt-hook: ignore build artifacts in git 

> apt-hook: refactor json hook messaging to be dry
> 
> The message construction code in the json hook
> was pretty repetitive. This cleans it up a bit
> and also removes some unnecessary explicit struct
> names in the tests.

> apt-hook: ignore go hook on some architectures
> 
> Golang doesn't support powerpc and riscv on older versions.
> This configures our Build-Depends and Makefile to gracefully
> handle scenarios where we can't build our go hook. We will
> simply not ship the go hook in these limited scenarios.
> 
> Fixes: #1603

> build: order build-depends alternatives newer first
> 
> When using alternatives in the Build-Depends section of
> debian/control, we should always put the newer version of
> a package first. This is because sbuild doesn't resolve
> alternatives by default and we want sbuild to work on the
> latest version of ubuntu with default settings.

> build: specify debianutils min version
> 
> This fixes a lintian error where we had an
> essential package that didn't specify version.



## Test Steps

1. Build on xenial through hirsute and make sure everything succeeds
2. Manually edit a go test to fail, then try to build and make sure the build fails

## Desired commit type::


## Checklist:
